feat: i18n framework

espeon opened the pull request.

espeon committed:
base i18n framework

espeon committed:
swap to i18next

espeon committed:
Use dropdown for language search, update translations

espeon committed:
Add update-related translations for es-ES and fix zh-TW app name

Commit:
extract to json -> ftl

Commit:
fix updates getting messed up

Commit:
update translations

Commit:
reorganize and compile i18n stuff

Commit:
try convert to zero-based styles for consistency

Commit:
regen lockfile

iameli reviewed (changes requested):
Let's chat about the workflow here, want to make sure it's easy as possible for devs + translators here. Ideally it wouldn't require a manual step in between devs adding a new translation string and seeing it pop up on the app

Random stuff:

1. Language settings don't seem to persist across refresh
2. Feeling like this should probably be a `components` thing rather than an `app` thing, assuming our users are gonna want to provide translations as well
3. We've got several copies of everything here, ideally even if we need it copied we'd have one set of `.ftl` files committed and everything else would get generated
4. `scripts/extract-i18n.js` seems to clobber the organization in the .ftl files, is the plan to not use it?

espeon commented on js/docs/src/content/docs/guides/localizing/for-devs.md:
```diff
@@ -0,0 +1,128 @@
+      components={{
+        supportLink: <a href="mailto:support@stream.place" />,
+        docsLink: <a href="/docs" />,
+        bold: <strong />,
```
yeah you're right it shouldn't. good point

Commit:
Move i18n setup to @streamplace/components

Commit:
it works on mobile (again) now

Commit:
Move genned files to public/locales, zh-hant

Commit:
build properly

Commit:
remove hans (we don't have any translations for that yet!)

for posterity:

"zh-Hans": { "code": "zh-Hans", "name": "Chinese (Simplified)",
"nativeName": "简体中文", "flag": "汉" },

Commit:
make example a little more react native friendly

Commit:
update lockfile again

iameli commented on js/components/src/i18n/i18next-config.ts:
```diff
@@ -0,0 +1,275 @@
+            translations = require("../../public/locales/fr-FR/messages.json");
+            break;
+          default:
+            throw new Error(`No static translation file for locale: ${locale}`);
```
I was about to tell you that this wouldn't work, 'cause the compiler isn't smart enough to not bundle these at build-time just because they're in the else block of `if (typeof window !== "undefined") {`. But... it totally is, they don't get bundled on web. Huh. Good job, Metro.

espeon replied:
I was surprised myself honestly

iameli commented on js/components/scripts/compile-translations.js:
```diff
@@ -0,0 +1,210 @@
+  }
+
+  return JSON.stringify(translations, null, 2);
+}
```
Surely there's a library for this so we don't have to hand-roll it? Looking at https://github.com/projectfluent/fluent.js

espeon replied:
hmm, looks right: iNaturalist does basically that here: https://github.com/inaturalist/iNaturalistReactNative/blob/f6b3ba1841e9c1c93f5530a68c0fa67c622fcd05/src/i18n/i18ncli/ftl.js#L175

iameli reviewed (changes requested):
Cool, getting there:
 
* Looks like the docs page isn't in the sidebar
* Suggested the use of a library for fluent.js stuff so we don't miss corner cases
* At least in dev, language selection seems to be not persistent; I'm getting English again after I refresh
* (Also in dev) it seems like I'm getting a FOUC on page load - settings page says `app-version` and whatnot for a moment until the translation loads. We can just do a `visibility: hidden` on everything until you get your localization, I reckon
* For a user adding translations, the dev flow doesn't totally work; you can run `i18n:watch` but that doesn't copy translations into `js/app/public` so you don't see the new translations in dev. It'd be nice if `i18n:watch` ran one command that built and copied the translations to where they need to be
* Need `nodemon` as a `devDependencies` in `components`, wasn't installed for me and it failed

I think there's room to improve the dev workflow further here but I don't wanna block getting this merged 👍

Commit:
readd I18nProvider to prevent FOUC

Commit:
save language to streamplace storage actually

Commit:
parse ftl using the official library, and convert to using namespaces

Commit:
copy to app when watching

espeon commented:
Should have fixed all those issues, except the docs page in the sidebar, I couldn't replicate that and I don't think it should have been affected by this at all

espeon commented:
Will refactor the selection off and then commit

Commit:
Move language over to dev settings for now

Commit:
executive decision （joking)
